VERSIONING.md

```diff
@@ -0,0 +1,220 @@
+
+          > Major version zero (0.y.z) is for initial development. Anything MAY
+          > change at any time. The public API SHOULD NOT be considered stable.
+
+        * Mature modules for which we guarantee a stable public API and telemetry will
+          be versioned with a major version greater than `v0`.
+        * Experimental modules will start their versioning at `v0.0.0` and will
+          increment their minor version when backwards incompatible changes are
+          released and increment their patch version when backwards compatible
+          changes are released.
+        * Stable contrib modules cannot depend on experimental modules from this
+          project.
+        * All stable contrib modules of the same major version with this project
+          will use the same entire version as this project.
+            * Stable modules may be released with an incremented minor or patch
+              version even though that module's code has not been changed. Instead
+              the only change that will have been included is to have updated that
+              modules dependency on this project's stable APIs.
+            * When an experimental module in contrib becomes stable a new stable
+              module version will be released and will include this now stable
+              module. The new stable module version will be an increment of the minor
+              version number and will be applied to all existing stable contrib
+              modules, this project's modules, and the newly stable module being
+              released.
+    * Contrib modules will be kept up to date with this project's releases.
+        * Due to the dependency contrib modules will implicitly have on this
+          project's modules the release of stable contrib modules to match the
+          released version number will be staggered after this project's release.
+          There is no explicit time guarantee for how long after this projects
+          release the contrib release will be. Effort should be made to keep them
+          as close in time as possible.
+        * No additional stable release in this project can be made until the
+          contrib repository has a matching stable release.
+        * No release can be made in the contrib repository after this project's
+          stable release except for a stable release of the contrib repository.
+* GitHub releases will be made for all releases.
+* Go modules will be made available at Go package mirrors.
+
+## Example Versioning Lifecycle
+
+To better understand the implementation of the above policy the following
+example is provided. This project is simplified to include only the following
+modules and their versions:
+
+* `collector`: `v0.28.0`
+* `collector/component`: `v0.28.0`
+* `collector/config`: `v0.28.0`
+* `collector/consumer`: `v0.28.0`
+* `collector/receiver`: `v0.28.0`
+* `collector/exporter`: `v0.28.0`
```

Comment:
Not sure if this is part of this spec or was decided elsewhere, but do we need so many separate modules? Could we have `collector/stable` (or maybe just `collector`) `collector/unstable`, and `collector/experimental`.

Experimental can depend on unstable and stable.
Unstable can depend on stable.
Stable can only depend on itself.

Reply:
The number and granularity of modules I think is still to be defined, this is provided as an example to work through how the policy could be applied if there were modules at this granularity.

I would be firmly opposed to using the import path to distinguish between `stable`, `unstable`, and `experimental` modules.  The version number conveys that information and keeping it out of the import path means that we do not require users to change their import paths everywhere when a module advances from `unstable` to `stable`.

Reply:
What about just a note/comment instead of having that in the package path?

Reply:
@Aneurysm9 
> The version number conveys that information and keeping it out of the import path means that we do not require users to change their import paths everywhere when a module advances from unstable to stable.

It's a sed command to rename imports so I'm not sure it's a terribly high burden. There may be other reasons this isn't the best idea but I'm not sure changing import paths should be the dealbreaker.

@bogdandrutu 
> What about just a note/comment instead of having that in the package path?

My concern there is that it won't be obvious which paths are stable/unstable/experimental without having to manually check who is using what all the time. For example as part of CI we would probably want to lint things to make sure forbidden APIs weren't being used. I guess we could probably automate this if we encode that information in a machine-readable way. ie, have a standardized structure to commenting on the stability of the package that a script can extract. Maybe inside `doc.go` have a `// OTEL-STABILITY: stable` comment or whatever. 

Or.. would it be crazy to do it with build tags? ie

`// +build otel-unstable` 

Slightly annoying to have to put it in every file (only for unstable and experimental, stable wouldn't need any build tags).

I think as long as we can easily develop some tooling to enforce the package comments I'd be ok with it though.

Reply:
To the extent that multiple modules will not be used in this repository and the single (non-command) module will be declared stable atomically when it is given a `v1.0.0` release, I don't believe any of this example is relevant and I am removing it from the next draft.